### PR TITLE
[Payload Inspector] Add more SiPixel inspection classes

### PR DESCRIPTION
**Pull request by mmusich**
#### PR description:

This PR introduces several improvement of the SiPixel Payload Inspector suite:
   * introduces inspection of the `SiPixelTemplate` Condition Format;
   * introduces new plot format types, `Phase1PixelMaps` and `PixelRegionContainers` in order to display cleaner Pixel Tracker Maps and values / comparison plots split by Layer (Barrel) / Ring (Endcap);
   * deploy the new plots for the `SiStripLorentzAngle` and `SiPixelGainCalibrationOffline(ForHLT)`;
   * migrate several classes to use the new template recommended after integration of https://github.com/cms-sw/cmssw/pull/29622;
   * deploy new unit tests and test scripts;

#### PR validation:

Relies on the existing and newly introduced unit tests. 
All the new plots have been tested also by hand; few examples are reproduced here: 

   *  `PixelRegionContainers` comparing two `SiPixelGainCalibrationOffline` payloads from two different tags
image

   * `Phase1PixelMaps` of the assume value of μH for one payload of `SiPixelTemplate`
image

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This PR is not a backport and no backport is intended.

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by ggovi** [comment]
+1

**Comment by silviodonato** [comment]
+1
I will merge this in order to have it in CMSSW_11_1_0.
@mmusich, could you please make another PR to remove "magic numbers" (you can use `constexpr`), to remove the commented code, and to replace `std::cout` with `edm::LogVerbatim` or something like that?

**Comment by mmusich** [comment]
> could you please make another PR to remove "magic numbers" (you can use constexpr), to remove the commented code, and to replace std::cout with edm::LogVerbatim or something like that?

@silviodonato sure, but please point to an example to either one of these. 

To clarify a bit in advance:
   *  `COUT` is already protected from the a compilation flag here:
https://github.com/cms-sw/cmssw/blob/5aef21d82888637a3daffcaf49579b56dd88bb73/CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h#L30-L35

   *  the "magic numbers" are mostly due to the geometry which is not accessible without the `EventSetup` (as in the Payload Inspector)

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp [hunk 1]
```diff
@@ -119,5 +120,29 @@ int main(int argc, char** argv) {
   histo16.process(connectionString, PI::mk_input(tag, end, end));
   std::cout << histo16.data() << std::endl;
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:121]: cout

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp [hunk 2]
```diff
@@ -119,5 +120,29 @@ int main(int argc, char** argv) {
...
 
+  boost::python::dict inputs;
+  inputs["SetLog"] = "True";  // sets to true, 1,True,Yes will work
+
+  SiPixelGainCalibrationOfflineGainsValuesBarrel histo17;
+  histo17.setInputParamValues(inputs);
+  histo17.process(connectionString, PI::mk_input(tag, end, end));
+  std::cout << histo17.data() << std::endl;
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:129]: cout

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp [hunk 3]
```diff
@@ -119,5 +120,29 @@ int main(int argc, char** argv) {
...
+
+  // SiPixelTemplates
+
+  tag = "SiPixelTemplateDBObject38Tv3_express";
+  start = boost::lexical_cast<unsigned long long>(326083);
+  end = boost::lexical_cast<unsigned long long>(326083);
+
+  std::cout << "## Exercising SiPixelTemplates plots " << std::endl;
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:137]: cout

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp [hunk 4]
```diff
@@ -119,5 +120,29 @@ int main(int argc, char** argv) {
...
+
+  SiPixelTemplateIDsBPixMap histo18;
+  histo18.process(connectionString, PI::mk_input(tag, end, end));
+  std::cout << histo18.data() << std::endl;
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:141]: cout

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp [hunk 5]
```diff
@@ -119,5 +120,29 @@ int main(int argc, char** argv) {
...
+
+  SiPixelTemplateLAFPixMap histo19;
+  histo19.process(connectionString, PI::mk_input(tag, end, end));
+  std::cout << histo19.data() << std::endl;
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:145]: cout

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+    for (unsigned i = 0; i < 4; ++i) {
+      histName = "barrel_layer_";
+
+      th2p = std::make_shared<TH2Poly>((histName + std::to_string(i + 1)).c_str(),
+                                       Form("PXBMap of %s - Layer %i", what, i + 1),
+                                       -15.0,
+                                       15.0,
+                                       0.0,
+                                       5.0);
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:43]: numbers

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    for (unsigned side = 1; side <= 2; ++side) {
+      for (unsigned disk = 1; disk <= 3; ++disk) {
+        histName = "forward_disk_";
+
+        th2p = std::make_shared<TH2Poly>((histName + std::to_string((side == 1 ? -(int(disk)) : (int)disk))).c_str(),
+                                         Form("PXFMap of %s - Side %i Disk %i", what, side, disk),
+                                         -15.0,
+                                         15.0,
+                                         -15.0,
+                                         15.0);
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:80]: numbers

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+        th2p->GetYaxis()->SetTitle("y [cm]");
+        //th2p->GetXaxis()->SetTitleOffset(0.09);
+        //th2p->GetYaxis()->SetTitleOffset(0.09);
+        th2p->SetStats(false);
+        th2p->SetOption(m_option);
+        pxfTh2PolyForward[currentHistoName].push_back(th2p);
+      }
+    }
+
+    th2p = std::make_shared<TH2Poly>("forward_summary", "PXFMap", -40.0, 50.0, -20.0, 90.0);
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:92]: numbers

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+      float vertX[] = {theVectX[0], theVectX[1], theVectX[2], theVectX[3], theVectX[4]};
+      float vertY[] = {(ladder - 1.0f), (ladder - 1.0f), (float)ladder, (float)ladder, (ladder - 1.0f)};
+
+      bins[id] = std::make_shared<TGraph>(5, vertX, vertY);
+      bins[id]->SetName(TString::Format("%u", id));
+
+      // Summary plot
+      for (unsigned k = 0; k < 5; ++k) {
+        vertX[k] += ((layer == 2 || layer == 3) ? 0.0f : -60.0f);
+        vertY[k] += ((layer > 2) ? 30.0f : 0.0f);
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:127]: numbers

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+      float vertX[] = {theVectX[0], theVectX[1], theVectX[2], theVectX[3]};
+      float vertY[] = {theVectY[0], theVectY[1], theVectY[2], theVectY[3]};
+
+      bins[id] = std::make_shared<TGraph>(4, vertX, vertY);
+      bins[id]->SetName(TString::Format("%u", id));
+
+      // Summary plot
+      for (unsigned k = 0; k < 4; ++k) {
+        vertX[k] += (float(side) - 1.5f) * 40.0f;
+        vertY[k] += (disk - 1) * 35.0f;
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:165]: numbers

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+  //============================================================================
+  void DrawBarrelMaps(const std::string& currentHistoName, TCanvas& canvas) {
+    canvas.Divide(2, 2);
+    for (int i = 1; i <= 4; i++) {
+      canvas.cd(i);
+      if (strcmp(m_option, "text") == 0) {
+        canvas.cd(i)->SetRightMargin(0.02);
+        pxbTh2PolyBarrel[currentHistoName].at(i - 1)->SetMarkerColor(kRed);
+      } else {
+        SiPixelPI::adjustCanvasMargins(canvas.cd(i), 0.07, 0.12, 0.10, 0.14);
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:232]: numbers

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+  //============================================================================
+  void DrawForwardMaps(const std::string& currentHistoName, TCanvas& canvas) {
+    canvas.Divide(3, 2);
+    for (int i = 1; i <= 6; i++) {
+      canvas.cd(i);
+      if (strcmp(m_option, "text") == 0) {
+        canvas.cd(i)->SetRightMargin(0.02);
+        pxfTh2PolyForward[currentHistoName].at(i - 1)->SetMarkerColor(kRed);
+      } else {
+        SiPixelPI::adjustCanvasMargins(canvas.cd(i), 0.07, 0.12, 0.10, 0.16);
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:247]: numbers

File: CondCore/SiPixelPlugins/interface/PixelRegionContainers.h [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+        TPaveStats* st = (TPaveStats*)plot.second->FindObject("stats");
+        if (st) {
+          st->SetTextSize(0.03);
+          st->SetLineColor(10);
+          if (plot.second->GetFillColor() != 0) {
+            st->SetTextColor(plot.second->GetFillColor());
+          } else {
+            st->SetTextColor(plot.second->GetLineColor());
+          }
+          SiPixelPI::adjustStats(st, 0.13, 0.85 - index * 0.08, 0.36, 0.93 - index * 0.08);
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/PixelRegionContainers.h:282]: numbers

File: CondCore/SiPixelPlugins/interface/SiPixelGainCalibHelper.h [hunk 14]
```diff
@@ -293,16 +306,362 @@ namespace gainCalibHelper {
...
+      if (setLog) {
+        myPlots.setLogScale();
+      }
+      myPlots.beautify(kBlue, 10);
+      myPlots.draw(canvas, isBarrel, "HIST");
+
+      TLegend legend = TLegend(0.45, 0.88, 0.91, 0.92);
+      legend.SetHeader(("hash: #bf{" + (std::get<1>(iov)) + "}").c_str(),
+                       "C");  // option "C" allows to center the header
+      //legend.AddEntry(h1.get(), ("IOV: " + std::to_string(std::get<0>(iov))).c_str(), "PL");
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/SiPixelGainCalibHelper.h:413]: commented line

File: CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h [hunk 15]
```diff
@@ -794,10 +800,10 @@ namespace SiPixelPI {
...
   // overloaded method: mask entire module
   /*--------------------------------------------------------------------*/
-  std::vector<std::pair<int, int> > maskedForwardRocsToBins(int ring, int blade, int panel, int disk)
+  std::vector<std::pair<int, int>> maskedForwardRocsToBins(int ring, int blade, int panel, int disk)
   /*--------------------------------------------------------------------*/
   {
-    std::vector<std::pair<int, int> > rocsToMask;
+    std::vector<std::pair<int, int>> rocsToMask;
 
     //int nblade_list[2] = {11, 17};
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h:808]: commented line

File: CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h [hunk 16]
```diff
@@ -828,11 +834,11 @@ namespace SiPixelPI {
...
   /*--------------------------------------------------------------------*/
-  std::vector<std::tuple<int, int, int> > maskedForwardRocsToBins(
+  std::vector<std::tuple<int, int, int>> maskedForwardRocsToBins(
       int ring, int blade, int panel, int disk, std::bitset<16> bad_rocs, bool isFlipped)
   /*--------------------------------------------------------------------*/
   {
-    std::vector<std::tuple<int, int, int> > rocsToMask;
+    std::vector<std::tuple<int, int, int>> rocsToMask;
 
     //int nblade_list[2] = {11, 17};
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h:843]: commented line

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+                                         Form("PXFMap of %s - Side %i Disk %i", what, side, disk),
+                                         -15.0,
+                                         15.0,
+                                         -15.0,
+                                         15.0);
+        th2p->SetFloat();
+        th2p->GetXaxis()->SetTitle("x [cm]");
+        th2p->GetYaxis()->SetTitle("y [cm]");
+        //th2p->GetXaxis()->SetTitleOffset(0.09);
+        //th2p->GetYaxis()->SetTitleOffset(0.09);
```
Comment (silviodonato) [inline CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:85]: commented line

**Review by silviodonato — COMMENTED** [review]
Here some examples. `std::cout` was in `test`, it is not problematic. About the constants, even if they are trivial, I think it is better to define a `constexpr`, especially if the same number is repeated in several points.

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp (hunk 1 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:121]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/9bdb0593f91bd7f5525e8a307db1e78f77568dfe

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp (hunk 2 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:129]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/9bdb0593f91bd7f5525e8a307db1e78f77568dfe

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp (hunk 3 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:137]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/9bdb0593f91bd7f5525e8a307db1e78f77568dfe

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp (hunk 4 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:141]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/9bdb0593f91bd7f5525e8a307db1e78f77568dfe

File: CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp (hunk 5 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/test/testSiPixelPayloadInspector.cpp:145]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/9bdb0593f91bd7f5525e8a307db1e78f77568dfe

File: CondCore/SiPixelPlugins/interface/SiPixelGainCalibHelper.h (hunk 14 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/interface/SiPixelGainCalibHelper.h:413]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/0384924a87726caddb979becbc8a7d6946a95813

File: CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h (hunk 15 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h:808]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/0384924a87726caddb979becbc8a7d6946a95813

File: CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h (hunk 16 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/interface/SiPixelPayloadInspectorHelper.h:843]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/0384924a87726caddb979becbc8a7d6946a95813

File: CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h (hunk 17 above)
Comment (mmusich) [reply CondCore/SiPixelPlugins/interface/Phase1PixelMaps.h:85]: addressed at https://github.com/cms-sw/cmssw/pull/29938/commits/0384924a87726caddb979becbc8a7d6946a95813

**Pull request merged**